config/management/genesis/src/validator_builder.rs

```diff
@@ -0,0 +1,404 @@
+        rng.fill_bytes(&mut validator_network_address_encryption_key);
+
+        Self {
+            root_key,
+            treasury_compliance_key,
+            validator_network_address_encryption_key,
+            validator_network_address_encryption_key_version: 0,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ValidatorBuilder {
+    config_directory: PathBuf,
+    num_validators: usize,
+    randomize_first_validator_ports: bool,
+    template: NodeConfig,
+}
+
+impl ValidatorBuilder {
+    pub fn new<T: Into<PathBuf>>(config_directory: T) -> Self {
+        Self {
+            config_directory: config_directory.into(),
+            num_validators: 1,
+            randomize_first_validator_ports: true,
+            template: NodeConfig::default_for_validator(),
+        }
+    }
+
+    pub fn randomize_first_validator_ports(mut self, value: bool) -> Self {
+        self.randomize_first_validator_ports = value;
+        self
+    }
+
+    pub fn num_validators(mut self, num_validators: usize) -> Self {
+        self.num_validators = num_validators;
+        self
+    }
+
+    pub fn template(mut self, template: NodeConfig) -> Self {
+        self.template = template;
+        self
+    }
+
+    pub fn build(mut self) -> Result<(RootKeys, Vec<ValidatorConfig>)> {
+        // TODO Pass this in
+        let mut rng = rand::rngs::OsRng;
+
+        // Canonicalize the config directory path
+        self.config_directory = self.config_directory.canonicalize()?;
+
+        // Generate chain root keys
+        let root_keys = RootKeys::generate(&mut rng);
+
+        // Generate and initialize Validator configs
+        let mut validators = (0..self.num_validators)
+            .map(|i| {
+                self.initialize_validator_config(
+                    i,
+                    &mut rng,
+                    root_keys.validator_network_address_encryption_key,
+                    root_keys.validator_network_address_encryption_key_version,
+                )
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        // Build genesis
+        let mut genesis_storage =
+            OnDiskStorage::new(self.config_directory.join("genesis-storage.json"));
+        let (genesis, waypoint) =
+            Self::genesis_ceremony(&mut genesis_storage, &root_keys, &validators)?;
+
+        // Insert Genesis and Waypoint into each validator
+        for validator in &mut validators {
+            validator.insert_genesis(&genesis)?;
+            validator.insert_waypoint(&waypoint)?;
+
+            // verify genesis
+            let validator_storage = Storage::from(validator.storage());
+            let output = verify_genesis(
+                StorageWrapper::new("validator", validator_storage),
+                Some(validator.config.execution.genesis_file_location.as_path()),
+            )?;
+
+            anyhow::ensure!(
+                output.split("match").count() == 5,
+                "Failed to verify genesis"
+            );
```

Comment:
Why is this 5?  Maybe a comment would help explain the magic number?

Reply:
I have zero clue, this is code copied from the old config builder

Reply:
because we expect there to be 4 matches! I think. I guess we could have just looked for that string 4 times.